File: src/tools/edit.rs
//! Edit tool - str_replace, insert, create

use crate::{BoxFuture, Tool, ToolResult};
use crate::tools::session;
use serde::Deserialize;
use serde_json::Value;
use tokio::fs;

#[derive(Debug, Clone, Deserialize)]
pub struct EditArgs {
    pub command: String,  // view, str_replace, insert, create
    pub path: String,
    // view
    #[serde(default)]
    pub view_range: Option<Vec<i64>>,
    // str_replace
    #[serde(default)]
    pub old_str: Option<String>,
    #[serde(default)]
    pub new_str: Option<String>,
    // insert
    #[serde(default)]
    pub insert_line: Option<i64>,
    #[serde(default)]
    pub insert_text: Option<String>,
    // create
    #[serde(default)]
    pub file_text: Option<String>,
    /// Execute in session sandbox via MCP
    #[serde(default)]
    pub session_id: Option<String>,
}

pub struct EditTool;

impl Tool for EditTool {
    fn name(&self) -> &'static str { "text_editor" }
    fn description(&self) -> &'static str { "Edit files with view/str_replace/insert/create" }
    
    fn schema(&self) -> Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "command": {"type": "string", "enum": ["view", "str_replace", "insert", "create"], "description": "Edit command"},
                "path": {"type": "string", "description": "File path"},
                "view_range": {"type": "array", "items": {"type": "integer"}, "description": "[start, end] lines for view"},
                "old_str": {"type": "string", "description": "Text to find (str_replace)"},
                "new_str": {"type": "string", "description": "Replacement text (str_replace)"},
                "insert_line": {"type": "integer", "description": "Line to insert after"},
                "insert_text": {"type": "string", "description": "Text to insert"},
                "file_text": {"type": "string", "description": "File content (create)"},
                "session_id": {"type": "string", "description": "Execute in session sandbox"}
            },
            "required": ["command", "path"]
        })
    }
    
    fn execute(&self, args: Value) -> BoxFuture<'_, ToolResult> {
        Box::pin(async move {
            let args: EditArgs = match serde_json::from_value(args.clone()) {
                Ok(a) => a,
                Err(e) => return ToolResult::err(format!("Invalid args: {e}")),
            };
            
            // Route to session if provided
            if let Some(session_id) = &args.session_id {
                let mut mcp_args = serde_json::json!({
                    "command": args.command,
                    "path": args.path,
                });
                if let Some(v) = &args.view_range { mcp_args["view_range"] = serde_json::json!(v); }
                if let Some(v) = &args.old_str { mcp_args["old_str"] = serde_json::json!(v); }
                if let Some(v) = &args.new_str { mcp_args["new_str"] = serde_json::json!(v); }
                if let Some(v) = &args.insert_line { mcp_args["insert_line"] = serde_json::json!(v); }
                if let Some(v) = &args.insert_text { mcp_args["insert_text"] = serde_json::json!(v); }
                if let Some(v) = &args.file_text { mcp_args["file_text"] = serde_json::json!(v); }
                return call_mcp_tool(session_id, "text_editor", mcp_args).await;
            }
            
            // Local execution
            match args.command.as_str() {
                "view" => view_file(&args.path, args.view_range).await,
                "str_replace" => str_replace(&args.path, args.old_str.as_deref().unwrap_or(""), args.new_str.as_deref().unwrap_or("")).await,
                "insert" => insert_at(&args.path, args.insert_line.unwrap_or(0), args.insert_text.as_deref().unwrap_or("")).await,
                "create" => create_file(&args.path, args.file_text.as_deref().unwrap_or("")).await,
                _ => ToolResult::err(format!("Unknown command: {}", args.command)),
            }
        })
    }
}

async fn call_mcp_tool(session_id: &str, tool_name: &str, args: Value) -> ToolResult {
    match session::call_tool_in_session(session_id, tool_name, args).await {
        Ok(result) => {
            let content = result.get("content")
                .and_then(|c| c.as_array())
                .and_then(|arr| arr.first())
                .and_then(|c| c.get("text"))
                .and_then(|t| t.as_str())
                .unwrap_or("");
            let is_error = result.get("isError").and_then(|e| e.as_bool()).unwrap_or(false);
            if is_error { ToolResult::err(content.to_string()) } else { ToolResult::ok(content.to_string()) }
        }
        Err(e) => ToolResult::err(e),
    }
}

async fn view_file(path: &str, range: Option<Vec<i64>>) -> ToolResult {
    let content = match fs::read_to_string(path).await {
        Ok(c) => c,
        Err(e) => return ToolResult::err(format!("Read failed: {e}")),
    };
    let lines: Vec<&str> = content.lines().collect();
    let (start, end) = match range {
        Some(r) if r.len() >= 2 => ((r[0].max(1) - 1) as usize, if r[1] == -1 { lines.len() } else { r[1] as usize }),
        Some(r) if r.len() == 1 => ((r[0].max(1) - 1) as usize, lines.len()),
        _ => (0, lines.len()),
    };
    let result: Vec<String> = lines[start..end.min(lines.len())]
        .iter().enumerate()
        .map(|(i, l)| format!("{:>6} | {}", start + i + 1, l))
        .collect();
    ToolResult::ok(result.join("\n"))
}

async fn str_replace(path: &str, old: &str, new: &str) -> ToolResult {
    let content = match fs::read_to_string(path).await {
        Ok(c) => c,
        Err(e) => return ToolResult::err(format!("Read failed: {e}")),
    };
    let count = content.matches(old).count();
    if count == 0 { return ToolResult::err("No match found for old_str"); }
    if count > 1 { return ToolResult::err(format!("Multiple matches ({count}). old_str must be unique.")); }
    let new_content = content.replace(old, new);
    if let Err(e) = fs::write(path, &new_content).await {
        return ToolResult::err(format!("Write failed: {e}"));
    }
    ToolResult::ok("Replaced successfully")
}

async fn insert_at(path: &str, line: i64, text: &str) -> ToolResult {
    let content = match fs::read_to_string(path).await {
        Ok(c) => c,
        Err(e) => return ToolResult::err(format!("Read failed: {e}")),
    };
    let mut lines: Vec<&str> = content.lines().collect();
    let idx = if line <= 0 { 0 } else { (line as usize).min(lines.len()) };
    for (i, new_line) in text.lines().enumerate() {
        lines.insert(idx + i, new_line);
    }
    if let Err(e) = fs::write(path, lines.join("\n")).await {
        return ToolResult::err(format!("Write failed: {e}"));
    }
    ToolResult::ok(format!("Inserted at line {}", line))
}

async fn create_file(path: &str, content: &str) -> ToolResult {
    if fs::metadata(path).await.is_ok() {
        return ToolResult::err("File exists. Use str_replace to modify.");
    }
    if let Some(parent) = std::path::Path::new(path).parent() {
        let _ = fs::create_dir_all(parent).await;
    }
    if let Err(e) = fs::write(path, content).await {
        return ToolResult::err(format!("Create failed: {e}"));
    }
    ToolResult::ok(format!("Created: {path}"))
}

File: runtime/tools/edit.go
package tools

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/dreamyang-liu/ash/runtime/events"
)

// EditTool provides file viewing and editing with one command-dispatched schema.
type EditTool struct{}

func (e *EditTool) Name() string { return "text_editor" }

func (e *EditTool) Description() string {
	return "View or edit files. write creates or overwrites the full file."
}

func (e *EditTool) Schema() map[string]any {
	return map[string]any{
		"type": "object",
		"properties": map[string]any{
			"command":     map[string]any{"type": "string", "enum": []string{"view", "str_replace", "insert", "write"}, "description": "Command to execute: view, str_replace, insert, or write"},
			"path":        map[string]any{"type": "string", "description": "File path"},
			"view_range":  map[string]any{"type": "array", "items": map[string]any{"type": "integer", "minimum": 1}, "minItems": 2, "maxItems": 2, "description": "[start, end] inclusive line range for view"},
			"old_str":     map[string]any{"type": "string", "description": "Text to find (str_replace)"},
			"new_str":     map[string]any{"type": "string", "description": "Replacement text (str_replace)"},
			"insert_line": map[string]any{"type": "integer", "minimum": 0, "description": "Line number to insert after. Use 0 to insert at the start."},
			"insert_text": map[string]any{"type": "string", "description": "Text to insert"},
			"file_text":   map[string]any{"type": "string", "description": "Full file content (write creates or overwrites)"},
		},
		"required": []string{"command", "path"},
		"oneOf": []map[string]any{
			{
				"properties": map[string]any{"command": map[string]any{"const": "view"}},
				"required":   []string{"command", "path"},
			},
			{
				"properties": map[string]any{"command": map[string]any{"const": "str_replace"}},
				"required":   []string{"command", "path", "old_str", "new_str"},
			},
			{
				"properties": map[string]any{"command": map[string]any{"const": "insert"}},
				"required":   []string{"command", "path", "insert_line", "insert_text"},
			},
			{
				"properties": map[string]any{"command": map[string]any{"const": "write"}},
				"required":   []string{"command", "path", "file_text"},
			},
		},
	}
}

func (e *EditTool) Execute(args map[string]any) Result {
	command, _ := args["command"].(string)
	path, _ := args["path"].(string)
	if path == "" {
		return Err("path is required")
	}

	switch command {
	case "view":
		return e.view(path, args)
	case "str_replace":
		return e.strReplace(path, args)
	case "insert":
		return e.insert(path, args)
	case "write":
		return e.write(path, args)
	default:
		return Err("unknown command: " + command)
	}
}

func (e *EditTool) view(path string, args map[string]any) Result {
	data, err := os.ReadFile(path)
	if err != nil {
		return Err(err.Error())
	}

	lines := strings.Split(string(data), "\n")
	start, end := 1, len(lines)

	if vr, ok := args["view_range"].([]any); ok && len(vr) == 2 {
		if s, ok := vr[0].(float64); ok {
			start = int(s)
		}
		if e, ok := vr[1].(float64); ok {
			end = int(e)
		}
	}

	if start < 1 {
		start = 1
	}
	if end > len(lines) {
		end = len(lines)
	}

	var b strings.Builder
	for i := start - 1; i < end; i++ {
		fmt.Fprintf(&b, "  %6d | %s\n", i+1, lines[i])
	}
	return Ok(b.String())
}

func (e *EditTool) strReplace(path string, args map[string]any) Result {
	oldStr, ok := args["old_str"].(string)
	if !ok || oldStr == "" {
		return Err("old_str is required for str_replace")
	}
	newStr, ok := args["new_str"].(string)
	if !ok {
		return Err("new_str is required for str_replace")
	}

	data, err := os.ReadFile(path)
	if err != nil {
		return Err(err.Error())
	}

	content := string(data)
	count := strings.Count(content, oldStr)

	if count == 0 {
		return Err("No match found")
	}
	if count > 1 {
		return Err(fmt.Sprintf("Multiple matches (%d). old_str must be unique.", count))
	}

	content = strings.Replace(content, oldStr, newStr, 1)
	if err := os.WriteFile(path, []byte(content), 0644); err != nil {
		return Err(err.Error())
	}

	events.Push("file_change", path, map[string]any{"path": path, "operation": "str_replace"})
	return Ok("OK")
}

func (e *EditTool) insert(path string, args map[string]any) Result {
	rawInsertLine, ok := args["insert_line"].(float64)
	if !ok {
		return Err("insert_line is required for insert")
	}
	insertLine := int(rawInsertLine)
	insertText, ok := args["insert_text"].(string)
	if !ok {
		return Err("insert_text is required for insert")
	}

	data, err := os.ReadFile(path)
	if err != nil {
		return Err(err.Error())
	}

	lines := strings.Split(string(data), "\n")
	if insertLine < 0 || insertLine > len(lines) {
		return Err(fmt.Sprintf("insert_line out of range: got %d, want 0..%d", insertLine, len(lines)))
	}
	newLines := strings.Split(insertText, "\n")

	result := make([]string, 0, len(lines)+len(newLines))
	result = append(result, lines[:insertLine]...)
	result = append(result, newLines...)
	result = append(result, lines[insertLine:]...)

	if err := os.WriteFile(path, []byte(strings.Join(result, "\n")), 0644); err != nil {
		return Err(err.Error())
	}

	events.Push("file_change", path, map[string]any{"path": path, "operation": "insert"})
	return Ok("OK")
}

func (e *EditTool) write(path string, args map[string]any) Result {
	fileText, ok := args["file_text"].(string)
	if !ok {
		return Err("file_text is required for write")
	}

	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
		return Err(err.Error())
	}
	if err := os.WriteFile(path, []byte(fileText), 0644); err != nil {
		return Err(err.Error())
	}

	events.Push("file_change", path, map[string]any{"path": path, "operation": "write"})
	return Ok("OK")
}
